Maps Update: Cards and Modals for Maps Components

juanjoseguva opened the pull request:
# Description

Updating Maps page to display as cards with modals, matching the other pages. 

UPDATE: 
 - This PR has been updated to also migrate over to RTKQuery for data fetching, and slices, for state management.
 - Fixes: #1197, #1145, #667 ,#1192, #1146 immutability check and serializable checks are both enabled
 
 2509 Update:
- Fixes: #1147


## Type of change

- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.github.io/developer/cla.html) and each author is listed in the Description section.

## Limitations

Saving functionality should be moved into the modals, and the save changes button under the cards should be deleted.

UPDATE: The migration maintains much of the prior functionality but, for simplicity, some of the maps logging was temporarily omitted. Logging as a whole should be carefully planned and implemented in a future PR for the omitted logs from maps, but fort he application as a whole.

juanjoseguva committed:
Changing maps page to cards with modals

juanjoseguva committed:
Adding messagesfor map modal

juanjoseguva committed:
Adding licence header

juanjoseguva committed:
Removing unused line

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,91 @@
+				{props.map.name} {props.isSubmitting ? '(Submitting)' : props.isEdited ? '(Edited)' : ''}
+			</div>
+			<div className="item-container">
+				<b><FormattedMessage id="map.displayable" /></b>
```
Other pages have a colon (:) after the descriptor on the card. This is also true on map edit.

Also, OED is not longer putting the item name in the descriptors (so no map).

juanjoseguva replied:
Resolved

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,91 @@
+				<b><FormattedMessage id="map.filename" /></b> {props.map.filename}
+			</div>
+			<div className="item-container">
+				<b><FormattedMessage id="note" /></b> {props.map.note}
```
Note should be limited to 30 characters as in other modals.

huss replied:
I am unclear on why this was resolved since my look at the code does not see this addressed. Am I missing something?

hazeltonbw replied:
I'm not sure why it's showing this old code here. In VSCode the latest commit has the changes to limit notes to 30 characters.

huss replied:
I see it too so it seems okay now.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,151 @@
+					</FormGroup>
+				</Form>
+				<div>
+					<Label><FormattedMessage id="map.filename" /></Label>
```
Please see conversion edit for how uneditable items should look.

huss replied:
From what I see this still does not look the way the mentioned page works. Am I missing something?

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,151 @@
+					<p>{props.map.filename}</p>
+					<Link to='/calibration' onClick={() => handleCalibrationSetting(CalibrationModeTypes.initiate)}>
+						<Button color='primary'>
+							<FormattedMessage id='map.upload.new.file' />
```
I think the name of this button needs to be clearer given the change to the modal and its location.

juanjoseguva replied:
Renamed button to a more descriptive phrase

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,151 @@
+							value={displayable.toString()}
+							onChange={e => setDisplayable(e.target.value === 'true')}
+						>
+							<option value="true">{intl.formatMessage({ id: 'map.is.displayable' })}</option>
```
This is done via an enum on other pages such as unit displayable. Also, I think this should mirror the bidirectional on conversion so yes/no with automatic coloring.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,151 @@
+					<FormGroup>
+						<Label for="mapCircleSize"><FormattedMessage id="map.circle.size" /></Label>
+						<Input
+							id="mapCircleSize"
```
This is a number and should be treated as such. It should also make sure it is > 0.

juanjoseguva replied:
Changed

huss replied:
This was changed but I think the way it is done does not conform to other uses and does not stop saving on invalid.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,151 @@
+				<FormattedMessage id="edit.map" />
+			</ModalHeader>
+			<ModalBody>
+				<Form>
```
I think I appreciate the ordering on this modal but it does not match the card. Maybe change card to match.

juanjoseguva replied:
Applied order change to card

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,91 @@
 
 interface MapViewProps {
-	// The ID of the map to be displayed
 	id: number;
```
This is a big picture comment that I am putting here. In a conversion to the Redux Toolkit, most props and all the containers should go away. (For maps some will likely remain for the new map/calibration until that is converted/updated. I have not tried but I think it will be okay to remove the others while leaving these.) These still exist and need to be removed. I suspect the issue with needing a save on the card page rather than each edit modal relates to this. See other pages for the general idea and this should mean you can pass just the needed map state (actually id is good enough but that is not what most pages do). This is going to lead to non-trivial changes in the code so I'm not going to do a full review until this is done. If you need help with this then work with your team and/or contact me.

juanjoseguva replied:
This has been addressed with Brandon's most recent commits

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,91 @@
+
+	useEffect(() => {
+		if (props.isEdited) {
+			//updateUnsavedChanges();
```
I think this is no used by OED at this time and probably should go from the page as it seems to be on other pages.

huss reviewed (changes requested):
~~Thanks to @juanjoseguva for doing this work. I've made some comments that will require non-trivial changes. Please let me know if anything is not clear or I can help.~~

Nireves333 committed:
Fix to allow saves in map modal

juanjoseguva committed:
Adding ": " to map card descriptors for formatting

juanjoseguva committed:
Removing unnecessary button

hazeltonbw committed:
Create and utilize new maps selector

hazeltonbw committed:
Class Component to Redux function component update

juanjoseguva committed:
Adding missing : on a map descriptor

juanjoseguva committed:
Removing item name from descriptor (so no map)

Commit:
Limiting map note to 30 characters.

Commit:
Treating circle size as a positive number

Commit:
Changing the order of descriptors on card to match the modal.

juanjoseguva committed:
Removing unused variable

Commit:
Changing the name of the button to upload a new map image file.

Commit:
Resolving syntax and jsdoc declaration requirements

Commit:
Resolving syntax and jsdoc declaration requirements

hazeltonbw committed:
use useAppDispatch for type safety

hazeltonbw committed:
Remove container props, utilize dispatch for redux actions

hazeltonbw committed:
Use the new component instead of old container

hazeltonbw committed:
Remove props used for container, removed isEdited and isSubmitting status as they are no longer relevant in the new modals

hazeltonbw committed:
Delete unused types and only import necessary functions from moment package

hazeltonbw committed:
Delete unused props & functions, add new memoized selector

juanjoseguva committed:
Treating map file name as an uneditable item

hazeltonbw commented:
~~All comments have been addressed and this PR is ready for review @huss.~~

hazeltonbw commented:
~~This PR includes some migrations of map components/containers to RTK as requested in issue #1145, however more components such as the calibration components need updated still.~~

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,72 @@
+	const handleShowModal = () => setShowEditModal(true);
+	const handleCloseModal = () => setShowEditModal(false);
+
+	const selectMapById = React.useMemo(makeSelectMapById, []);
```
My knowledge is not good enough to know if useMemo is needed. Having said this, I'm unsure what this is doing/needed for. It may just take me more time but if you have any hints that might help.

ChrisMart21 replied:
What's the rationale for using a factory method here for the selector? Reselect generally doesn't require this anymore for multiple instances ever since their introduced weakMap memoization.

 See  https://reselect.js.org/introduction/v5-summary#breaking-changes

hazeltonbw replied:
This is selecting the map by ID. I used [unique selector instances](https://redux.js.org/usage/deriving-data-selectors#creating-unique-selector-instances) (useMemo) and a [selector factory](https://redux.js.org/usage/deriving-data-selectors#selector-factories) as it seemed to be the best option at the time. If there is a better way to do this please let me know and I will correct it. Thanks

ChrisMart21 replied:
Looking more closely at the selectors, this should not require a memoized createSelector . All it's doing is extracting the selected id from the map state. Since there is no data manipulation, or derivation it will always return a 'stable reference' to the map state entity. So a simple selector will suffice.

```
export const selectMapById= (state: RootState,id: number) => state.maps.byMapId[id];
```

P.S. Side note, even if this did need memoization, there was a newer update that essentially gets rid of the need for the factorySelector functions + useMemo() combo.
[WeakMapMemoize](https://reselect.js.org/api/weakMapMemoize) is now the baseline default 'memoize' function. Meaning that one instance used across multiple components will be memoized properly!

hazeltonbw replied:
Thank you for the explanation, changes have been made on commit [811dc66](https://github.com/OpenEnergyDashboard/OED/pull/1314/commits/811dc6693629b4bef6f78de78981095a93f66e0d)

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,72 @@
+	const handleCloseModal = () => setShowEditModal(false);
+
+	const selectMapById = React.useMemo(makeSelectMapById, []);
+	const map = useSelector(state => selectMapById(state, mapID));
```
This is a shift to a newer way but I don't think it goes all the way to useAppSelector. This may need to be a future issue since it may not be resolved in this PR.

hazeltonbw replied:
Changes have been made on commit [811dc66](https://github.com/OpenEnergyDashboard/OED/pull/1314/commits/811dc6693629b4bef6f78de78981095a93f66e0d)

huss replied:
For the record, I think still uses props and a container that needs to be updated.

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,72 @@
+			</div>
+			<div className="item-container">
+				<b><FormattedMessage id="map.displayable" /></b>
+				<span style={{ color: map.displayable ? 'green' : 'red' }}>
```
When I look at the unitView page, for example, it does not do the coloring this way but via TrueFalseType. This would be consistent and centralize the coloring and make it look the same.

Nireves333 replied:
Addressed in commit [51a220f](https://github.com/OpenEnergyDashboard/OED/pull/1314/commits/51a220f23a762ee9ed372cce910ea7915b292766)

I believe this is what you meant, the TrueFalseType as used in the unitView does not touch any coloring of text. Let me know if this should be modified differently.

huss replied:
The not displayable is not working correctly.

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,72 @@
+				<b><FormattedMessage id="map.circle.size" /></b> {map.circleSize}
+			</div>
+			<div className="item-container">
+				<b><FormattedMessage id="note" /></b> {map.note}
```
Note should be limited to 30 characters as on other modals.

huss replied:
This does limit on the card but also limits on edit which is not what is wanted. The display on card is limited not the actual note.

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,72 @@
+			</div>
+			<div className="item-container">
+				<b><FormattedMessage id="map.modified.date" /></b>
+				{parseZone(map.modifiedDate, undefined, true).format('dddd, MMM DD, YYYY hh:mm a')}
```
Not your doing but I don't think this will propewrly internationalize.

Nireves333 replied:
Addressed in commit [51a220f](https://github.com/OpenEnergyDashboard/OED/pull/1314/commits/51a220f23a762ee9ed372cce910ea7915b292766)

Marked as TODO

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,72 @@
+			</div>
+			<div className="item-container">
+				<b><FormattedMessage id="map.calibration" /></b>
+				<span style={{ color: map.origin && map.opposite ? 'black' : 'gray' }}>
```
An advantage of the old code is it only checked the origin/opposite once where here it is done twice. Nice if function or combined somehow.

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,72 @@
-					<Button style={this.styleToggleBtn()} color='primary'>
-						<FormattedMessage id='map.calibrate' />
+			</div>
+			{hasToken() && (
```
I think you can only get to this page if you are an admin. Is this needed?

juanjoseguva replied:
You're right, only admins can access this page. Will remove the token check.

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -2,436 +2,72 @@
-						<FormattedMessage id='map.calibrate' />
+			</div>
+			{hasToken() && (
+				<div className="edit-btn">
```
This differs from unit. I'm curious if this is deliberate or if standardizing makes sense?

juanjoseguva replied:
Aligned with unit

huss commented on src/client/app/components/maps/CreateMapModalComponent.tsx:
```diff
@@ -0,0 +1,80 @@
+
+/**
+ * Defines the create map modal form
+ * @param root0
```
The JSDoc is off and it is an unusual param values. Could you help me understand?

hazeltonbw replied:
This will get wiped in future changes.

juanjoseguva committed:
Handling displayable with an enum instead of a boolean and fixing a typo in the translations.

huss commented on src/client/app/components/maps/CreateMapModalComponent.tsx:
```diff
@@ -0,0 +1,80 @@
+ * @param root0.createNewMap
+ * @returns Map create element
+ */
+function CreateMapModalComponent({ show, handleClose, createNewMap }: CreateMapModalProps) {
```
Other create modals do not pass props. Are these essential or does a change need to happen (maybe outside this PR)?

hazeltonbw replied:
This will be future work, I have left a TODO note.

huss commented on src/client/app/components/maps/CreateMapModalComponent.tsx:
```diff
@@ -0,0 +1,80 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
I'm unclear as what this file is used for. When I click create map it goes to the old page. I cannot find it used anywhere. Is it for the future, unneeded or am I missing something? If it is future then a clear TODO/comment is needed.

hazeltonbw replied:
This is work for the future, as there was not enough time to finish it given our time constraints. I have left TODOs throughout the necessary pieces of code.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+
+interface EditMapModalProps {
+	show: boolean;
+	handleClose: () => void;
```
This is probably a note for the future. The view passes the id but here is passed the map state object. At some point OED needs to standardize.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+}
+
+const EditMapModalComponent: React.FC<EditMapModalProps> = ({ show, handleClose, map}) => {
+	const dispatch: AppDispatch = useAppDispatch();
```
Note for future to move to newer way. Similar to another comment.

hazeltonbw replied:
Noted for future work.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+
+const EditMapModalComponent: React.FC<EditMapModalProps> = ({ show, handleClose, map}) => {
+	const dispatch: AppDispatch = useAppDispatch();
+	const [nameInput, setNameInput] = useState(map.name);
```
I'm unclear on why the note could be empty if this is editing an existing note. Can the note be null or something?

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+		handleClose();
+	};
+
+	const handleDelete = () => {
```
This is not being done as is the case, for example, with unit. It should be the same.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+			handleClose();
+		}
+	};
+
```
Maybe for future TODO and in issue. Most modals use centralized functions for changes and not one for each item. This means you don't dispatch changes until save.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+		handleClose();
+	};
+
+	const toggleCircleEdit = () => {
```
This sort of check is now done with input checks on the form in other modals.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+						<Input
+							id='mapDisplayable'
+							name='mapDisplayable'
+							type='select'
```
I think it would be clear/consistent to have displayable as a boolean. Then this check is not needed and it looks as other pages. See edit meter.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+					<Input
+						id='mapFilename'
+						name='mapFilename'
+						type='text'
```
Another origin/opposite check where one function that is reused would be good.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+							value={noteInput}
+							onChange={e => setNoteInput(e.target.value.slice(0,30))}
+						/>
+					</FormGroup>
```
The cards show the label/name on one line. I think that is clearer. However done it would be good to be consistent.

When I click on new file nothing seems to happen but a message shows in the terminal/log for actions started.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+				<div>
+					<Label><FormattedMessage id="map.filename" /></Label>
+					<Input
+						id='mapFilename'
```
As above for file, standardize with card view.

I also want to say that I personally find the spacing to be confusing. I think the one line for title/info will help but also putting a space before this would help separate it from file.

huss replied:
The next separator resolves the spacing issue but not the placement of items together.

huss commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,156 @@
+						name='mapFilename'
+						type='text'
+						defaultValue={map.filename}
+						disabled>
```
Calibration seems to have same issue as file in nothing happens

huss commented on src/client/app/redux/actions/map.ts:
```diff
@@ -2,24 +2,20 @@
 
-import {ActionType, Dispatch, GetState, Thunk} from '../../types/redux/actions';
+import { ActionType, Dispatch, GetState, Thunk } from '../../types/redux/actions';
 import * as t from '../../types/redux/map';
```
OED is going away from import *. Having said that, I think a number of the imported items will go away when the new Redux way is used (for example, actions).

huss commented on src/client/app/redux/actions/map.ts:
```diff
@@ -2,24 +2,20 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
```
For the record, this file needs changes to not use State directly, change to later Redux usage, and maybe more. This is likely a future work.

hazeltonbw replied:
Noted for future work

huss commented on src/client/app/redux/reducers/maps.ts:
```diff
@@ -6,7 +6,6 @@ import { MapMetadata, MapsAction, MapState } from '../../types/redux/map';
 import { ActionType } from '../../types/redux/actions';
```
For the record, another file that needs changes for newer Redux. It also should get away from thinking of multiple edited maps, etc.

hazeltonbw replied:
Noted for future work

huss commented on src/client/app/redux/selectors/maps.ts:
```diff
@@ -0,0 +1,21 @@
+import { RootState } from 'store';
+import { createSelector } from '@reduxjs/toolkit';
+
+export const selectMapState = (state: RootState) => state.maps;
```
I don't know if getting RootState is what is now normally done so this may go away in the update to Redux usage.

Nireves333 replied:
I believe this may have been addressed in commit [811dc66](https://github.com/OpenEnergyDashboard/OED/pull/1314/commits/811dc6693629b4bef6f78de78981095a93f66e0d)

huss replied:
I'm sorry but I still see state being directly handled. Do you see otherwise?

hazeltonbw replied:
I'm not sure what you mean by state being directly handled. I see the use of createAppSelector() in other selector files, would you like me to use this instead?

hazeltonbw replied:
I have made changes in commit [d270242](https://github.com/OpenEnergyDashboard/OED/pull/1314/commits/d270242dbf65ef6dce23cefc65ce95fb3610b163) that I believe solves this issue.

ChrisMart21 replied:
He may be referring to the fact that these selectors are not using the same patterns of accessing data as meters, groups, etc.. Additionally many existing selectors, when possible, are defined  in slices/ exported. There aren't many using 
(state:Root state)=> state.select.my.thing;
The difference here is that maps isn't using RTKQ, or rtk slices, so I think this is fine and will change once the data fetching also migrates over where it can use the entityAdapter pattern like other apiData. 

@huss correct me if I'm mistaken

huss replied:
Yes, this is an older way that works but should be updated. It isn't part of the issue linked so it did not have to happen but I wanted the developers to know. I am noting this since some other state changes were made. More and more I'm leaving comments on PRs about items that need to happen (future changes, documentation, etc.)

huss commented on src/client/app/translations/data.ts:
```diff
@@ -795,27 +799,27 @@ const LocaleTranslationData = {
 		"map.is.calibrated": "Étalonnage terminé",
 		"map.is.deleted": "Carte supprimée de la base de données",
-		"map.is.displayable": "Affichage activé",
+		"map.is.displayable": "Affichage activé: ",
```
Only language where : was added.

juanjoseguva replied:
Good catch! This was an unnecessary edit.

huss commented:
> This PR includes some migrations of map components/containers to RTK as requested in issue #1145, however more components such as the calibration components need updated still.

~~Please either open a new issue or unlink the issue in the description since more work is needed.~~

huss reviewed (changes requested):
~~Thanks to the team for the updated code. I have made a number of comments. Given the time in the semester, we may want to chat to see what should get done and what needs documentation/issue(s) for the future and if/how this PR might be finalized. Having said this, the team is welcome to work on comments, esp. the more straightforward ones, as you see fit.~~

~~I also should have noted that these is a merge conflict.~~

juanjoseguva committed:
Bug squashing

juanjoseguva committed:
Aligning modal call function names with other components and limiting note length

hazeltonbw committed:
Update code to use simple selector

hazeltonbw committed:
Put modal state where it belongs

juanjoseguva committed:
Merge branch 'development' into Maps-Update

juanjoseguva committed:
Reverting to treating map displayability variable as a boolean instead of enum

Nireves333 committed:
Add coloring to map Display and add TODO for date internationalization

hazeltonbw committed:
Use memoized createAppSelector instead of directly handling state in selectors.

hazeltonbw committed:
Fix linting errors

hazeltonbw committed:
Leave TODO notes for future updates

hazeltonbw commented:
~~Comments have been addressed. A few items weren't able to get resolved due to time constraints (migrating legacy Redux to RTK). These items have been marked with TODO comments in the code. Several items were consolidated into one TODO comment as the necessary changes to the code will solve more than one comment. Changes to the modals have been updated and this PR is ready for review.~~

ChrisMart21 commented on src/client/app/components/maps/MapsDetailComponent.tsx:
```diff
@@ -5,114 +5,54 @@
+ */
+// TODO: Migrate to RTK
+export default function MapsDetailComponent() {
+	const dispatch: AppDispatch = useAppDispatch();
```
It is a strong reccomendation from the redux team to try avoid importing from store whenever possible. Typescript will infer this type from the useAppDispatch ReturnValue type, so import is not needed here.

ChrisMart21 commented on src/client/app/components/maps/EditMapModalComponent.tsx:
```diff
@@ -0,0 +1,160 @@
+import { editMapDetails, submitEditedMap, removeMap, setCalibration } from '../../redux/actions/map';
+import { showErrorNotification } from '../../utils/notifications';
+import { useAppDispatch } from '../../redux/reduxHooks';
+import { AppDispatch } from 'store';
```
Same as other comment, unneeded import AppDispatch from store.

ChrisMart21 commented on src/client/app/components/maps/MapCalibrationComponent.tsx:
```diff
@@ -6,8 +6,9 @@ import * as React from 'react';
 import MapCalibrationInfoDisplayContainer from '../../containers/maps/MapCalibrationInfoDisplayContainer';
 import MapCalibrationInitiateContainer from '../../containers/maps/MapCalibrationInitiateContainer';
-import MapsDetailContainer from '../../containers/maps/MapsDetailContainer';
+//import MapsDetailContainer from '../../containers/maps/MapsDetailContainer';
```
No longer need this comment?

ChrisMart21 commented:
~~From a fresh docker instance, I was able to create a map, didn't complete calibration, and saved immediately. Then successfully edited, and saved map details like name, circleSize, however clicking on calibrate from EditMapModalComponent does not navigate to the calibration page properly, Was not able to enter calibration for unCalibrated map.~~

huss committed:
Merge remote-tracking branch 'origin/development' into pr/juanjoseguva/1314

huss committed:
fix linting

ChrisMart21 committed:
Initial RTKQ Maps

huss commented:
~~This is a good start toward moving maps to cards and some newer Redux usage. However, there are remaining comments to address and more work to be done. Since the team working on this is done (please let me know if it is otherwise), it will need to be continued/completed by others. Given this, I am converting it to draft state so future work can utilize this work.~~

ChrisMart21 commented:
I'd like to continue the work of this PR if that's alright.

huss commented:
> I'd like to continue the work of this PR if that's alright.

Thanks to @ChrisMart21 as knowing someone is pushing this forward would be great. Let me know if I can help in any way.

ChrisMart21 committed:
Remove Maps from RootState
- maps no longer a state property,
- old State type deleted, only use RootState now
- admin pages untested, graph seems okay.
- needs testing.

ChrisMart21 committed:
Comments

ChrisMart21 committed:
More Maps Changes

ChrisMart21 committed:
ReEnable Immutable check

ChrisMart21 commented:
@huss 
Some updates have been made including. 
 - legacy 'maps' property has been removed from RootState and is longer reliant on older State/Actions. 
 - Maps Data Fetching now using RTK-Query.
 - Removed Image from redux state, instead process image metaData upon data fetch, and utilize mapSource where possible.
 - DevModeCheck : 'immutabilityCheck' seems safe to re-enable.
 
Have yet to test, so its likely revisions are needed.

ChrisMart21 committed:
Bug Fix Revert  Change.

ChrisMart21 committed:
Hot Fix Changes.

ChrisMart21 committed:
Fix Imports

ChrisMart21 committed:
MapChartSelect Crash Fix

ChrisMart21 committed:
Login Invalidation updated.

ChrisMart21 committed:
Cleanup Legacy Redux where applicable
 - Legacy Actions/Creators deleted,
 - unused actions deleted
 - unused types deleted
 - maps files purged.
 - comments.

ChrisMart21 committed:
Merge branch 'development' into Maps-Update

ChrisMart21 committed:
Fix Merge Issues / Missing MPL headers

ChrisMart21 committed:
Update Root Reducer, Sanitize Devtools

ChrisMart21 committed:
Address Unserializable Values In State/Actions.
 - queryTimeIntervalString
 - rangeSliderIntervalString
 - barDuration
 - mapsBarDuration
 - compareTimeIntervalString
 - Re-Enable DevModeChecks

huss commented:
~~Wanted to note that I started to merge this with development to remove conflicts. There are some files that need careful merging due to the barDuration vs mapDuration (not longer separated by graphic type) and the map circle size. I plan to do this but am putting it off for now to move reviewing code forward.~~

Rakesh-Ranga-Buram commented on src/client/app/components/MapChartComponent.tsx:
```diff
@@ -34,6 +32,8 @@ import { AreaUnitType, getAreaUnitConversion } from '../utils/getAreaUnitConvers
 import getGraphColor from '../utils/getGraphColor';
 import translate from '../utils/translate';
```
I'm migrating translate to useTranslate as part of #1354 . Can you please migrate to useTranslate() from componentHooks.ts

huss committed:
Merge remote-tracking branch 'origin/development' into doMerge

This is a generally working version but more testing
is needed. It is showing serialization warnings.
More changes are expected during review and for other
changes since this work was done.

huss committed:
Change map edit to use yes/no in displayable menu

huss committed:
Remove unneeded items

- Remove MapState as no longer used.
- Remove console.log.

huss committed:
fix uppercase of en maps for pages menu

huss committed:
first step in fixing state so serializable

I made a mistake in the merge of development and put non-serializable items into state.
This had been fixed in the PR and I undid it. Also, new items from development needed
fixing. This first step corrects for rangeSliderInterval.

huss committed:
fix serializable for initialXAxisRange

huss committed:
fix serializable for queryTimeInterval

huss committed:
fix serializable for duration & compareTimeInterval

huss committed:
fix serializable for shiftTimeInterval

huss committed:
fix serializable for updateComparePeriod

huss committed:
various fixes

- Try to get all state update to only accept serializable arguments.
This involves a number of the update functions in graphSlice.ts.
The corresponding uses are also updated.
- Fix selectSliderRangeInterval to provide  an interval.
- Export raw readings still not done since had issue.
- Only moderately tested so more changes possible.

huss committed:
serializable & 2 bug fixes

- make raw export serializable for time.
- Fix bug where raw export failed if no meters.
- Fix bug in raw export route where param checking logic
was incorrect so failed.

huss committed:
add comment on checking store

huss committed:
working & look changes for initiate map calibration

huss committed:
fix up map pages

- Formatting & wording on display page.
- Convert from log to Toast msgs.
- Control when can save calibration.
- Rename some translation keys.

huss committed:
remove :  after map identifier if no chang

huss committed:
reset clicked point + other

- Resets the clicked point after each successful submit.
- Removes console msg for mode.
- Some commenting.

huss committed:
change note length on map card

This only puts  ... if the note is too long. It also shortens to 25 characters so fits on one line of card.

huss committed:
return to maps page +

- After calibration, the user is returned to the card map overview page.
This also happens on edit but that does not change anything.
- Improve some msgs.
- Misc cleanup.

huss committed:
fix non-serializable for chart links

- The result of useSearchParams() was not serializable. To fix this
it is converted to a string before dispatching and put back
in the slice functions.
- The original code used URLSearchParams for the name of the
result from useSearchParams but that conflicted with the name
of the class so it was changed.

huss committed:
map delete confirmation to modal

- This matches usage in other files.
- The delete for unit had button color changed to danger.
Also changed to quotes around unit name.

huss commented:
I must begin by apologizing to @juanjoseguva, @hazeltonbw & @Nireves333 who did the original work on this PR and to @ChrisMart21 who did substantial updates. It is inappropriate that this work sat for so long and I take complete responsibility for that happening. It is no way a reflection upon the work done.

I have begun serious work to prepare this PR for merging. My recent push, at the highest level, does the following.

- Merges the latest development and resolves the conflicts.
- Changes the text and UI feel in a number of places to be consistent with the current code in OED. Some changes will need to wait until the calibration is fully integrated into the newer ways OED does operations.
- Addressed misc. items for small changes.
- Made code so there are (hopefully) no non-serializable state used with React/Redux. Some changes were my fault from the merging of development which removed some needed changes. Others addressed more state issues. I suspect (but am not certain) that more warnings were added from React/Redux about values passed to the Toolkit, etc. so those are now addressed.

There is some more worked needed to finish up the PR. I plan to do that but it may take a little while as my attention will be drawn to other matters. If anyone wants to work on this then let me know and I can provide my list of items that I know of.

Finally, if anyone has thoughts on my changes then they are more than welcome.

huss committed:
fix misspelled word

huss committed:
remove unused CreateMapModalComponent

huss committed:
add min/max form feedback on circle size

huss committed:
translations for MapChartComponent

huss committed:
translations +

- Unrelated to PR but missing translations noticed in other
graphic pages.
- I small formatting change.

huss committed:
remove calibration from map view

huss committed:
fix spacing around displayable + text

- There was no space before the map displayable in the view.
It is somewhat a hack but move to second position to fix.
Due to coloring changing className.
- Change map pages to use displayable like other pages.
- Remove unused translations for map displayable.

huss committed:
change how unsaved changes handled +

- Removes syncing of values to global map state including removing debounce.
- Removes popup on bad circle size since already in form feedback.
This removes the onBlur.
- Adds unsaved warning to page similar to other pages.
- Changes cancel to discard as other pages. It clears edits on discard.
- Makes save dependent on having a change.
- Added tooltip (help page needed).
- Comments and TODO as this does not completely decide
if using local state as other pages or global state.
- Remove toggle on ModalHeader to be consistent.
- Spacing before button.

huss committed:
fix incorrect import

huss committed:
map duration chart link fix

aduques commented:
As discussed with @huss, I will continue on the work in this PR. Many of the changes have been worked on, and the necessary work involves resolving existing merge conflicts and addressing possible security changes. As I have noted from our discussion,
"need to change security changes, when merges come in, the merges include serializeable map data"
